File: cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"

#include <algorithm> // To use std::reverse

namespace secret_handshake {

/**
 * @brief This implements the secret handshake algorithm in the
 * exercise. It takes an integer and returns a vector of strings
 * that represent the secret handshake.
 * 
 * The actions for each number place are:
 *
 *  00001 = wink
 *  00010 = double blink
 *  00100 = close your eyes
 *  01000 = jump
 *  10000 = Reverse the order of the operations in the secret handshake.
 * 
 * @param n is the integer to be converted to a secret handshake. 
 * @return vector<string> is the secret handshake.
 */
vector<string> commands(int n) {
    vector<string> result;
    if (n & 1) result.push_back("wink");
    if (n & 2) result.push_back("double blink");
    if (n & 4) result.push_back("close your eyes");
    if (n & 8) result.push_back("jump");
    if (n & 16) reverse(result.begin(), result.end());
    return result;
}

}  // namespace secret_handshake

File: haskell/secret-handshake/src/SecretHandshake.hs
{-# LANGUAGE BinaryLiterals #-}

{- | Secret Handshake module for the exercism.io Haskell Track
-}

module SecretHandshake (handshake) where

import Data.Bits ( Bits((.&.)) )

{- | handshake function returns a list of strings based on the secret handshake rules.

    The task is to convert a number to a list of strings based on the following rules:
    1. The least significant bit is a wink
    2. The second least significant bit is a double blink
    3. The third least significant bit is a close your eyes
    4. The fourth least significant bit is a jump
    5. The fifth least significant bit is a reverse the order of the operations in the secret handshake.

    We solved this task by recursively calling the handshake function with the next relevant bits
    until we reach the end of the bits. We then reverse the list if the fifth bit is set, which is
    possible because it is the first match we check for.
-}
handshake :: Int -> [String]
handshake n 
    | relevantBits >= 0b10000 = reverse (handshake ((.&.) relevantBits 0b01111))
    | relevantBits >= 0b01000 = handshake ((.&.) relevantBits 0b00111) ++ ["jump"]
    | relevantBits >= 0b00100 = handshake ((.&.) relevantBits 0b00011) ++ ["close your eyes"]
    | relevantBits >= 0b00010 = handshake ((.&.) relevantBits 0b00001) ++ ["double blink"]
    | relevantBits == 0b00001 = ["wink"]
    | otherwise = []
    where relevantBits = (.&.) n 0b11111